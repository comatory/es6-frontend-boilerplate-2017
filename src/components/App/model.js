import { Record } from 'immutable';

export default class AppModel extends Record({
  background: '#FFF',
  requestStatus: null,
}) {}
