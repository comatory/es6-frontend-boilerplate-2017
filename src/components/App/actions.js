export const CHANGE_BACKGROUND = 'CHANGE_BACKGROUND';
export const CALL_ASYNC_ACTION_END = 'CALL_ASYNC_ACTION_END';
export const CALL_ASYNC_ACTION_START = 'CALL_ASYNC_ACTION_START';

export function changeBackground() {
  return {
    type: CHANGE_BACKGROUND,
  };
}

export function callAsyncAction() {
  return (dispatch) => {
    dispatch({
      type: CALL_ASYNC_ACTION_START,
    });
    setTimeout(
      () => {
        dispatch({
          type: CALL_ASYNC_ACTION_END,
        });
      }, 2000,
    );
  };
}
