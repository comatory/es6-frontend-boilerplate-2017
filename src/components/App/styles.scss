@import '../../style/colors.scss';
@import '../../style/containers.scss';
@import '../../style/inputs.scss';

.app_body {
  width: 80%;
  max-width: 1280px;
  margin: 20px auto;
  background-color: $primary;
  font-family: sans-serif;
  border-radius: $container-border-radius;
  min-width: 360px;
}

.app_wrapper {
  padding: 30px;
  :global(h1) {
    margin-top: 0;
    color: $light;
    letter-spacing: 0.03em;
  }
}

.wrapper {
  margin: 10px 0px;
}

.inner_wrapper {
  background-color: #FFF;
  border-radius: $container-border-radius;
  padding: 10px;
}

.button {
  @include button;
}