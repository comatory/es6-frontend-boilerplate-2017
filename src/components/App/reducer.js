import AppModel from './model';
import { CHANGE_BACKGROUND, CALL_ASYNC_ACTION_START, CALL_ASYNC_ACTION_END } from './actions';

export default function appReducer(state = new AppModel(), action) {
  switch (action.type) {
    case CHANGE_BACKGROUND:
      return state.set('background', `#${Math.random().toString(16).slice(2, 8).toUpperCase()}`);
    case CALL_ASYNC_ACTION_START:
      return state.set('requestStatus', 'Fake request is running...');
    case CALL_ASYNC_ACTION_END:
      return state.set('requestStatus', 'Fake request finished!');
    default:
      return state;
  }
}
