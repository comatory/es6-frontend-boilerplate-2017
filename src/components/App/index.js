import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import styles from './styles.scss';
import { changeBackground, callAsyncAction } from './actions';

export class App extends React.Component {
  static propTypes = {
    changeBackground: PropTypes.func.isRequired,
    callAsyncAction: PropTypes.func.isRequired,
    background: PropTypes.string.isRequired,
    requestStatus: PropTypes.string,
  }

  static defaultProps = {
    requestStatus: null,
  }

  render() {
    const { changeBackground, background, requestStatus, callAsyncAction } = this.props;
    return (
      <div className={styles.app_body}>
        <div className={styles.app_wrapper}>
          <h1>Sample app</h1>
          <div className={styles.wrapper}>
            <p>
              This is sample React component.
              Put your components in <code>/src/components/</code> directory.
            </p>
          </div>
          <div className={styles.wrapper}>
            <div className={styles.inner_wrapper} style={{ backgroundColor: background }}>
              <button className={styles.button} onClick={changeBackground}>Change color</button>
              <p>
                <b>Current color:</b>{background}
              </p>
              <button className={styles.button} onClick={callAsyncAction}>Call async action</button>
              { requestStatus && <p>{requestStatus}</p>}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  const { background, requestStatus } = state.app;
  return {
    background,
    requestStatus,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    changeBackground,
    callAsyncAction,
  }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
