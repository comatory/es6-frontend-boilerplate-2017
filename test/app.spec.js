/* eslint no-undef: [0], "react/react-in-jsx-scope": [0] */
import { App } from '../src/components/App';

describe('App', () => {
  const props = {
    changeBackground: () => {},
    callAsyncAction: () => {},
    background: '#FFF',
  };
  const wrapper = shallow(<App {...props} />);

  it('should render App component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
