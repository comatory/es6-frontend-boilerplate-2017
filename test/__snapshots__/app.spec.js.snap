// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render App component 1`] = `
<div
  className="app_body"
>
  <div
    className="app_wrapper"
  >
    <h1>
      Sample app
    </h1>
    <div
      className="wrapper"
    >
      <p>
        This is sample React component. Put your components in 
        <code>
          /src/components/
        </code>
         directory.
      </p>
    </div>
    <div
      className="wrapper"
    >
      <div
        className="inner_wrapper"
        style={
          Object {
            "backgroundColor": "#FFF",
          }
        }
      >
        <button
          className="button"
          onClick={[Function]}
        >
          Change color
        </button>
        <p>
          <b>
            Current color:
          </b>
          #FFF
        </p>
        <button
          className="button"
          onClick={[Function]}
        >
          Call async action
        </button>
      </div>
    </div>
  </div>
</div>
`;
